Mathlib/Analysis/Normed/Ring/SmoothingSeminorm.lean

```diff
@@ -0,0 +1,625 @@
+  use 1
+  intro n hn
+  simp only [smoothingSeminormSeq]
+  have hn0 : (n : ℝ) ≠ 0 := cast_ne_zero.mpr (one_le_iff_ne_zero.mp hn)
+  rw [hx n hn, ← rpow_natCast, ← rpow_mul (apply_nonneg μ _), mul_one_div_cancel hn0, rpow_one]
+
+/-- If `μ 1 ≤ 1` and `∀ y : R, μ (x * y) = μ x * μ y`, then `smoothingFun μ x = μ x`. -/
+theorem smoothingFun_apply_of_isMul (hμ1 : μ 1 ≤ 1) {x : R}
+    (hx : ∀ y : R, μ (x * y) = μ x * μ y) : smoothingFun μ x = μ x := by
+  apply tendsto_nhds_unique_of_eventuallyEq (tendsto_smoothingFun_of_map_one_le_one μ hμ1 x)
+    tendsto_const_nhds
+  simp only [EventuallyEq, eventually_atTop, ge_iff_le]
+  use 1
+  intro n hn
+  simp only [smoothingSeminormSeq]
+  by_cases hx0 : μ x = 0
+  · have hxn : μ (x ^ n) = 0 := by
+      apply le_antisymm _ (apply_nonneg μ _)
+      apply le_trans (map_pow_le_pow μ x (one_le_iff_ne_zero.mp hn))
+      rw [hx0, zero_pow (pos_iff_ne_zero.mp hn)]
+    rw [hx0, hxn, zero_rpow (one_div_cast_ne_zero (one_le_iff_ne_zero.mp hn))]
+  · have h1 : μ 1 = 1 := by rw [← mul_right_inj' hx0, ← hx 1, mul_one, mul_one]
+    have hn0 : (n : ℝ) ≠ 0 := cast_ne_zero.mpr (_root_.ne_of_gt (lt_of_lt_of_le zero_lt_one hn))
+    rw [← mul_one (x ^ n), pow_isMul_of_isMul μ hx, ← rpow_natCast, h1, mul_one, ←
+      rpow_mul (apply_nonneg μ _), mul_one_div_cancel hn0, rpow_one]
+
+/-- If `μ 1 ≤ 1`, `μ` is nonarchimedean, and `∀ y : R, μ (x * y) = μ x * μ y`, then
+  `smoothingSeminorm μ hμ1 hna x = μ x`. -/
+theorem smoothingSeminorm_apply_of_is_mul (hμ1 : μ 1 ≤ 1) (hna : IsNonarchimedean μ) {x : R}
+    (hx : ∀ y : R, μ (x * y) = μ x * μ y) : smoothingSeminorm μ hμ1 hna x = μ x :=
+  smoothingFun_apply_of_isMul μ hμ1 hx
+
+/-- If `μ 1 ≤ 1`, and `x` is multiplicative for `μ`, then it is multiplicative for
+  `smoothingFun`. -/
+theorem smoothingFun_of_isMul (hμ1 : μ 1 ≤ 1) {x : R} (hx : ∀ y : R, μ (x * y) = μ x * μ y)
+    (y : R) : smoothingFun μ (x * y) = smoothingFun μ x * smoothingFun μ y := by
+  have hlim : Tendsto (fun n => μ x * smoothingSeminormSeq μ y n) atTop
+      (𝓝 (smoothingFun μ x * smoothingFun μ y)) := by
+    rw [smoothingFun_apply_of_isMul μ hμ1 hx]
+    exact Tendsto.const_mul _ (tendsto_smoothingFun_of_map_one_le_one μ hμ1 y)
+  apply tendsto_nhds_unique_of_eventuallyEq (tendsto_smoothingFun_of_map_one_le_one μ hμ1 (x * y))
+    hlim
+  simp only [EventuallyEq, eventually_atTop, ge_iff_le]
+  use 1
+  intro n hn1
+  have hn0 : (n : ℝ) ≠ 0 := Nat.cast_ne_zero.mpr (_root_.ne_of_gt (lt_of_lt_of_le zero_lt_one hn1))
+  simp only [smoothingSeminormSeq]
+  rw [mul_pow, pow_isMul_of_isMul μ hx, mul_rpow (pow_nonneg (apply_nonneg μ _) _)
+    (apply_nonneg μ _), ← rpow_natCast, ← rpow_mul (apply_nonneg μ _), mul_one_div_cancel hn0,
+    rpow_one]
+
+/-- If `μ 1 ≤ 1`, `μ` is nonarchimedean, and `x` is multiplicative for `μ`, then `x` is
+  multiplicative for `smoothingSeminorm`. -/
+theorem smoothingSeminorm_of_mul (hμ1 : μ 1 ≤ 1) (hna : IsNonarchimedean μ) {x : R}
```

Comment:
Should this be `of_isMul` (and perhaps even something different if you agree with my comment about this naming in the `Basic.lean` file)?

Reply:
I have renamed this to `smoothingFun_apply_of_map_mul_eq_mul`. However, I am a bit worried that this makes it seem like the map is multiplicative, while the hypothesis is only that `x` is multiplicative with respect to that map.